### PR TITLE
Search cache: Force monitor to run every N IngestRun() invocations

### PR DESCRIPTION
**Pull request by mdittmer**
This change uses a channel to synchronize running a memory monitor check every N `IngestRun()` invocation on the index. The monitor trusts the index to synchronize on every invocation (which is not ideal, but works in practice).

### CODE REVIEW

**Comment by lukebjerring** [comment]
#1000 :tada:

File: api/query/cache/monitor/monitor.go [hunk 1]
```diff
@@ -114,19 +116,36 @@ func (m *indexMonitor) Start() error {
...
-		if heapBytes > m.maxHeapBytes {
-			m.logger.Errorf("Out of memory: %d > %d", heapBytes, m.maxHeapBytes)
-			m.idx.EvictRuns(m.percent)
-		} else {
-			m.logger.Debugf("Monitor: %d heap-allocated bytes OK", heapBytes)
+		timer := make(chan bool, 1)
+		go func() {
+			time.Sleep(m.interval)
+			timer <- true
+		}()
```
Comment (Hexcles) [inline api/query/cache/monitor/monitor.go]: I think you'd need to extract this channel and goroutine out of the infinite loop and add an infinite loop within the function; i.e., have a single, long-living goroutine that writes to the same `timer` channel periodically.

Imagine if you have runs coming in such a fast rate that eviction is never triggered by the timer, you'd end up with a whole bunch of unnecessary goroutines sleeping. Besides, `timer` is created in each iteration but never closed, and I'm not sure if this would confuse the GC.

File: api/query/cache/monitor/monitor.go [hunk 2]
```diff
@@ -114,19 +116,36 @@ func (m *indexMonitor) Start() error {
...
 			return errStopped
 		}
 
-		heapBytes := m.rt.GetHeapBytes()
-		if heapBytes > m.maxHeapBytes {
-			m.logger.Errorf("Out of memory: %d > %d", heapBytes, m.maxHeapBytes)
-			m.idx.EvictRuns(m.percent)
-		} else {
-			m.logger.Debugf("Monitor: %d heap-allocated bytes OK", heapBytes)
+		timer := make(chan bool, 1)
```
Comment (Hexcles) [inline api/query/cache/monitor/monitor.go]: No need to specify the capacity 1.

File: api/query/cache/monitor/monitor.go [hunk 3]
```diff
@@ -114,19 +116,36 @@ func (m *indexMonitor) Start() error {
...
+		go func() {
+			time.Sleep(m.interval)
+			timer <- true
+		}()
+
+		for {
+			select {
+			case <-timer:
+				m.check()
```
Comment (Hexcles) [inline api/query/cache/monitor/monitor.go]: Move `m.check()` (and the other one in the next `case`) out of this `for` loop.

File: api/query/cache/monitor/monitor.go [hunk 4]
```diff
@@ -175,13 +195,23 @@ func (m *indexMonitor) start() error {
 	return nil
 }
 
+func (m *indexMonitor) check() {
+	heapBytes := m.rt.GetHeapBytes()
+	if heapBytes > m.maxHeapBytes {
+		m.logger.Errorf("Out of memory: %d > %d", heapBytes, m.maxHeapBytes)
```
Comment (Hexcles) [inline api/query/cache/monitor/monitor.go]: This is not part of the PR, but I think we shouldn't use `Errorf` here as this is expected to happen eventually; and perhaps we should change the wording of the message to something like "%d memory allocated, exceeding the threshold %d" to avoid confusion with actual, fatal OOM.

File: api/query/cache/monitor/monitor.go (hunk 1 above)
Comment (mdittmer) [reply api/query/cache/monitor/monitor.go]: I don't think there's a GC issue. There are no references to `timer` after this goroutine exits.

The trouble with a separate goroutine is that it doesn't start its timeout after the last `check()`, it starts it independently. That means the monitor may run more frequently than intended.

File: api/query/cache/monitor/monitor.go (hunk 2 above)
Comment (mdittmer) [reply api/query/cache/monitor/monitor.go]: The capacity of 1 ensures that a goroutine that writes to `timer` can exit (and be cleaned up) immediately, rather than waiting for the `select` to choose `timer`.

File: api/query/cache/monitor/monitor.go (hunk 3 above)
Comment (mdittmer) [reply api/query/cache/monitor/monitor.go]: The `m.check()` in the other case is conditional. That's why I don't have it outside the `for` loop.

File: api/query/cache/monitor/monitor.go (hunk 4 above)
Comment (mdittmer) [reply api/query/cache/monitor/monitor.go]: Sounds good. We can make this a `Warningf` and soften the wording.

**Comment by Hexcles** [comment]
Please wait until #1005 is merged and rebase.
